flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java

```diff
@@ -118,6 +117,7 @@ public static Builder forRowData(DataStream<RowData> input) {
 
   public static class Builder {
     private DataStream<RowData> rowDataInput = null;
+    private SingleOutputStreamOperator<RowData> mappedRowDataInput = null;
```

Comment:
I'd prefer not introducing another `DataStream<RowData>` named `mappedRowDataInput` into this Builder,  because that could easily confuse people:  which `DataStream` should I use to chain the following operators ?  Let's just keep the only one `rowDataInput` even if we've chained a `MapFunction`,  we could still add the `uid` and `name` lazily to the `Map` operator ?

Reply:
Note that `mappedRowDataInput` is intentionally defined as `SingleOutputStreamOperator` (not `DataStream `), because `DataStream` doesn't expose `uid` and `name` method and hence we can't lazily add them to the `Map` operator with `DataStream<RowData> rowDataInput`.

I agree that this is a little confusing and less than ideal. Another option is to change Builder to a generic class like `Builder<T>`. but I don't like that either and hence used the current implementation. I am very open to other alternatives.

Reply:
I tried to initialize the input stream lazily as the following because I think it could make the source `rowDataInput` stream very clear, it's the only one stream that we will need to chain the following operator.

```diff
diff --git a/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java b/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java
index bda6667cd..a6be26b31 100644
--- a/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java
+++ b/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java
@@ -24,6 +24,7 @@ import java.io.UncheckedIOException;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
+import java.util.function.Function;
 import org.apache.flink.api.common.functions.MapFunction;
 import org.apache.flink.api.common.typeinfo.TypeInformation;
 import org.apache.flink.api.common.typeinfo.Types;
@@ -117,7 +118,7 @@ public class FlinkSink {
 
   public static class Builder {
     private DataStream<RowData> rowDataInput = null;
-    private SingleOutputStreamOperator<RowData> mappedRowDataInput = null;
+    private Function<String, DataStream<RowData>> inputCreator = null;
     private TableLoader tableLoader;
     private Table table;
     private TableSchema tableSchema;
@@ -136,9 +137,17 @@ public class FlinkSink {
     }
 
     private <T> Builder forMapperOutputType(DataStream<T> input,
-                                        MapFunction<T, RowData> mapper,
-                                        TypeInformation<RowData> outputType) {
-      this.mappedRowDataInput = input.map(mapper, outputType);
+                                            MapFunction<T, RowData> mapper,
+                                            TypeInformation<RowData> outputType) {
+      this.inputCreator = newUidPrefix -> {
+        if (newUidPrefix != null) {
+          return input.map(mapper, outputType)
+              .name(operatorName(newUidPrefix))
+              .uid(newUidPrefix + "-mapper");
+        } else {
+          return input.map(mapper, outputType);
+        }
+      };
       return this;
     }
 
@@ -241,8 +250,11 @@ public class FlinkSink {
     }
 
     public DataStreamSink<RowData> build() {
-      Preconditions.checkArgument(rowDataInput != null || mappedRowDataInput != null,
-          "Please use forRowData() or forMapperOutputType() to initialize the input DataStream.");
+      Preconditions.checkArgument(rowDataInput == null || inputCreator == null,
+          "Use forRowData() or forRow() to initialize the input DataStream only once.");
+      Preconditions.checkArgument(rowDataInput != null || inputCreator != null,
+          "Please use forRowData() or forRow() to initialize the input DataStream.");
+      rowDataInput = rowDataInput == null ? inputCreator.apply(uidPrefix) : rowDataInput;
       Preconditions.checkNotNull(tableLoader, "Table loader shouldn't be null");
 
       if (table == null) {
@@ -254,17 +266,6 @@ public class FlinkSink {
         }
       }
 
-      // set name and uid for mappedRowDataInput if needed
-      if (mappedRowDataInput != null) {
-        if (uidPrefix != null) {
-          rowDataInput = mappedRowDataInput
-              .name(operatorName(uidPrefix))
-              .uid(uidPrefix + "-mapper");
-        } else {
-          rowDataInput = mappedRowDataInput;
-        }
-      }
-
       // Convert the requested flink table schema to flink row type.
       RowType flinkRowType = toFlinkRowType(table.schema(), tableSchema);
```

Reply:
great idea. switched to this lazy ini mode